### ISSUE TITLE
Feature Request: Add support for envoy:ext_authz

### ISSUE THREAD

**Issue by tsraza** [pusher/oauth2_proxy#219]
Envoy proxy has authorization filter that expects a 200 response (plus useful headers), instead of a redirect to an upstream.

## Expected Behavior

Let envoy handle the redirect

## Current Behavior

The request gets forwarded to the upstream.

The only way to use with envoy is to pass the request directly to oauth2_proxy, rather than using the *ext_authz* filter.

## Possible Solution

Add a configuration option (boolean) to respond with status 200 when authorized, 403 when not (as opposed to 302).

## Steps to Reproduce (for bugs)

n/a

## Context

Envoy is an alternative to nginx for network stuff.

[external authorization architecture](https://www.envoyproxy.io/docs/envoy/v1.11.0/intro/arch_overview/security/ext_authz_filter#arch-overview-ext-authz)

## Your Environment

docker

**Comment by JoelSpeed** [comment]
Hey @tsraza,

We already have an endpoint that returns `202` and `401` responses depending on whether the user is authenticated or not, would that suffice? Could you try the `ext_authz` filter pointing at the `/oauth2/auth`  endpoint on your instance of the proxy?

**Comment by tsraza** [comment]
Thanks @JoelSpeed .  It gives a 404 either way if an upstream is not configured.

Even with an upstream configured, I think the only way to configure the `ext_authz` filter to hit `/oauth2/auth` is to prefix it to every request.  So the callback urls become `/oauth2/auth/oauth2/callback`.

Using ext_authz and setting the upstream to a backend service yields a timeout, even though I can see the backend is being hit with the right headers.

Setting the upstream back to envoy creates a neverending loop.

**Comment by tsraza** [comment]
@JoelSpeed I got it work with the `ext_authz` filter by hacking the handler and status codes in `oauthproxy.go`:

#### do not proxy (could not switch on upstream config given only `p`)

```go
func (p *OAuthProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
	switch path := req.URL.Path; {
	case path == p.RobotsPath:
		p.RobotsTxt(rw)
	case path == p.PingPath:
		p.PingPage(rw)
	case p.IsWhitelistedRequest(req):
		p.serveMux.ServeHTTP(rw, req)
	case path == p.SignInPath:
		p.SignIn(rw, req)
	case path == p.SignOutPath:
		p.SignOut(rw, req)
	case path == p.OAuthStartPath:
		p.OAuthStart(rw, req)
	case path == p.OAuthCallbackPath:
		p.OAuthCallback(rw, req)
	case path == p.AuthOnlyPath:
		p.AuthenticateOnly(rw, req)
	default:
		p.AuthenticateOnly(rw, req)
		// p.Proxy(rw, req)
	}
}
```

#### return 200 instead of 202

```go
func (p *OAuthProxy) AuthenticateOnly(rw http.ResponseWriter, req *http.Request) {
	status := p.Authenticate(rw, req)
	if status == http.StatusAccepted {
		rw.WriteHeader(http.StatusOK)
		// rw.WriteHeader(http.StatusAccepted)
	} else {
		http.Error(rw, "unauthorized request", http.StatusUnauthorized)
	}
}
```

**Comment by JoelSpeed** [comment]
Is this issue resolved by #265 ?

**Comment by tsraza** [comment]
Yes it is.  it works great
